### PR TITLE
Migrate to procps-ng 4.0

### PR DESCRIPTION
**Pull request by berarma**
I've tried to make the changes needed for procps-ng 4.0. This is the first time I touch any procps related code. I'm not sure how good are the changes. Needs testing.

### CODE REVIEW

File: compat/process-list.hpp [hunk 1]
```diff
@@ -131,33 +131,35 @@ static QStringList get_all_executable_names()
...
+            QString tmp(p_cmdline[0]);
             const int idx = std::max(tmp.lastIndexOf('\\'), tmp.lastIndexOf('/'));
             tmp = tmp.mid(idx == -1 ? 0 : idx+1);
             ret.append(tmp);
         }
-        freeproc(procs[i]);
     }
-    free(procs);
+
+    procps_pids_unref(&info);
```
Comment (sthalik) [inline compat/process-list.hpp:161]: - Incorrect code style (curly brace must be on separate line, etc)
- Don't malloc twice in the loop
- On my Arch Linux it has a lot of fake elements with `"-"` as the contents.

Run the code below and see if it still works for you. Try uncommenting the `qDebug()` statements to be sure.

```c++
    enum pids_item items[] = { PIDS_ID_PID, PIDS_CMD, PIDS_CMDLINE_V };
    
    while ((stack = procps_pids_get(info, PIDS_FETCH_TASKS_ONLY)))
    {
        enum rel_items { rel_pid, rel_cmd, rel_cmdline };
        struct pids_info *info = NULL;
        struct pids_stack *stack;
        QString tmp; tmp.reserve(64);

        procps_pids_new(&info, items, 3);

        while ((stack = procps_pids_get(info, PIDS_FETCH_TASKS_ONLY)))
        {
            char  **p_cmdline = PIDS_VAL(rel_cmdline, strv,  stack, info);

            // note, wine sets argv[0] so no parsing like in OSX case
            if (p_cmdline && p_cmdline[0] && p_cmdline[0][0] &&
                !(p_cmdline[0][0] == '-' && !p_cmdline[0][1]))
            {
                tmp = QString{p_cmdline[0]};
                const int idx = std::max(tmp.lastIndexOf('\\'), tmp.lastIndexOf('/'));
                tmp = tmp.mid(idx == -1 ? 0 : idx+1);
                //qDebug() << "procps" << tmp;
                ret.append(tmp);
            }
    }
    //qDebug() << "-- procps end";

    procps_pids_unref(&info);
```

**Review by sthalik — CHANGES_REQUESTED** [review]

File: compat/process-list.hpp (hunk 1 above)
Comment (berarma) [reply compat/process-list.hpp]: I've used your code removing a duplicated while statement at the start.

I'm testing on Debian 12 and there are no dashes, only valid commands. Could it be some security feature?

Thanks!

**Comment by sthalik** [comment]
Merged, thanks.

**Comment by sthalik** [comment]
In the future this code should use freelists with no mallocs at all but I don't have time right now. If it ain't broke don't fix it.

**Comment by sthalik** [comment]
I can't find `libproc2` for Ubuntu `jammy` so for now I have to support both versions. Will fix it today.

**Comment by berarma** [comment]
> I can't find `libproc2` for Ubuntu `jammy` so for now I have to support both versions. Will fix it today.

Yes, it seems Lunar is the first one that ships libproc2.

I would have liked to support both versions but I haven't enough experience with CMake/C++.

Thanks!

**Comment by berarma** [comment]
> @berarma please test either `master` or [e050725](https://github.com/opentrack/opentrack/commit/e0507257a41c29fa2e9ac28f36470023a72c39c6). Thanks.

It's working with libproc2 on Debian 12.

**Pull request merged**